WIP/ENH Gam 2744 rebased3

josef-pkt opened the pull request:
continuation of #5296 after rebase

commits for outside gam/penalized

- BUG/REF: fix GLM cov_params in fit_gradient, see #5316 
- REF: GLM protect against hessian inversion failure, use rtol in a test
- BUG: RegressionResults attach use_t if not None, see #5331
- BUG: correct GLM.estimate_scale using deviance
- ENH/REF: contrast results: string for float df_num, df_denom
- fix contrast summary formatting
- ENH: base LikelihoodModelResults.wald_test add df_constraints kwd

josef-pkt commented:
One main improvement to fmin alpha search is to add start_params to the underlying fit_pirls, e.g. start at previous (memoized) params.
For ex post inspection it would also be useful to log/store the params and aic/gcv at each evaluation during the fmin search.

josef-pkt commented:
trying to reverse engineer mgcv b-splines knot selection in one example.
`gam_a = gam(city.mpg ~ fuel + drive + s(weight,bs="bs",k=12) + s(hp,bs="bs",k=10), data = mpg)`

```
$knots
 [1]  624.3700  911.3873 1198.4047 1485.4220 1772.4393 2059.4567 2346.4740
 [8] 2633.4913 2920.5087 3207.5260 3494.5433 3781.5607 4068.5780 4355.5953
[15] 4642.6127 4929.6300
```
 
there are no knots at the min and max of the data.
it looks like the smallest and largest knot are extended by roughly 1/3 (0.335) of the distance max - min

```
# (largest_knot - data.max) / (data.max - data.min)
(4929.6300 - 4066) / (4066 - 1488) = 0.335

ext = (4066 - 1488) * 0.335  # (data.max - data.min) * 0.335
knots = np.linspace(1488 - ext, 4066 + ext, 16)
```

16 knots (k=12) are implied by the specification in statsmodels and R.

except it doesn't work if I increase the required number of knots 
e.g. if k=13, the extreme knots are only 0.3016 outside of the data extremes.

I don't remember having seen much of an explanation for the knot selection in mgcv.

weird, I thought bs chooses interior knots by quantiles, but in this example the knots are equal spaced

```
> kn = gam_a$smooth[[1]]$knots
> kn[2:17] - kn[1:16]
 [1] 287.0173 287.0173 287.0173 287.0173 287.0173 287.0173 287.0173 287.0173
 [9] 287.0173 287.0173 287.0173 287.0173 287.0173 287.0173 287.0173       NA
```

josef-pkt commented:
bsplines in mgcv use `contrast='center'`, i.e. they are not local but linearly transformed

spotchecking:
`bs = BSplines(x_spline, df=[12, 10], degree=[3, 3], variable_names=['weight', 'hp'],   
                     constraints='center', include_intercept=True)`

gives the same exog as the corresponding mgcv function when I force mgcv to use the same knots as statsmodels.

`gam_a = gam(city.mpg ~ fuel + drive + s(weight,bs="bs", k=12) + s(hp,bs="bs",k=10), data = mpg, knots=c(knots_w, knots_h))`

where knots in mgcv need to be a list of dataframes, AFAICS
```
sm_knots_w = c(1488., 1488., 1488., 1488., 1953.22222222, 2118.77777778, 2275., 2383.88888889, 2515.55555556, 2757.33333333, 3016.44444444, 3208.11111111, 4066., 4066., 4066., 4066.)
sm_knots_hp = c(48.0, 48.0, 48.0, 48.0, 68.0, 73.0, 88.0, 101.0, 116.0, 152.28571428571428, 288.0, 288.0, 288.0, 288.0)
knots_w <- data.frame(weight=sm_knots_w)
knots_h <- data.frame(hp=sm_knots_hp)
```

josef-pkt commented:
more trouble with b-splines
Our current implementation uses deriv2 at observations for penalty matrix. (comments above and in issue ?). This doesn't match up with mgcv. In the book Woods describes evaluating the deriv2 at one or several points within the knot intervals weighted by the knot distance.
The O-splines article uses a simplex integral approximation at knots and knot interval midpoints, but has it only for cubic b-splines, AFAICS.

I guess I need to add an option for the evaluation points for deriv2 for the penalty matrix sooner than I hoped to need it.
or I just use a scipy integrator for arrays, and match mgcv just approximately.

josef-pkt commented:
qsturng test `TestPsturng.test_100_random_values` failed
https://travis-ci.org/statsmodels/statsmodels/jobs/452642821
otherwise only lint errors (and a dev errorl)

josef-pkt commented:
Notes unit test and results exported form mgcv

- edf attribute in mgcv results is twice, I guess because I am adding gam and summary(gam) attributes to the same results Bunch class. needs renaming for unique attribute names

josef-pkt commented:
unit test for Poisson:
- sqrt in weights in pirls was wrong (mentioned in an earlier comment)
- after fixing this, pirls and newton give the same result as mgcv for given alpha, aic, edf also agree with mgcv
- estimated alpha/pen_weights with both aic and gcv seem to be too small, not enough smoothing based on wiggles in plot.
- I'm still not sure whether working resids should be used in ccpr plot for Poisson or cases with nonlinear link

based on reading code during debugging:

- comparing `GAM._fit_pirls` with `GLM._fit_irls`: there is a `n_trials` in weights computation missing, so Binomial with n_trials > 1 is most likely wrong

josef-pkt commented:
too good to be true?

selecting alpha by aic gives essentially the same result as mgcv UBRE 
They are asymptotically equivalent, monotone or linear transformation, but I didn't expect it too match so well.

I added a new model method `select_penweight` that currently defaults to basinhopping, nelder-mead and aic.
(bonus: `_fit_pirls` starts with previous params in search history)

```
aa, fr, hist = res1p.model.select_penweight()
aa
array([  8.28398074e+09,   1.46279930e+07])

alpha # mgcv
array([  8.28398928e+09,   1.46282076e+07])

1 / s_scale * sp / 2  # mgcv 
array([  8.28398928e+09,   1.46282076e+07])
```

josef-pkt commented:
AFAICS, the previous change of setting `include_intercept = True` is not the bug. The problem comes from the setup of the simulated data. The perfect or strong collinearity comes from using linspace to create the x sample points, i.e. x and x**2 have overlapping B-spline basis because of the polynomials of order 3 encompased by the splines
```
    x1 = np.linspace(-3, 3, n)
    x2 = np.linspace(0, 1, n)**2
```

changing x2 to `x2 = 10 / (1 + np.linspace(0, 1, n))` reduces concurvity strongly enough that pirls and bfgs agree. eigenvalues of exog are large enough.

mgcv has a `concurvity` function as diagnostic tool
 https://stat.ethz.ch/R-manual/R-devel/library/mgcv/html/concurvity.html

josef-pkt commented on statsmodels/gam/gam_penalties.py:
```diff
@@ -0,0 +1,209 @@
+        if weights is None:
+            # weights should hanve total length as params
+            # but it can also be scalar in individual
+            self.weights = [1. for _ in range(self.k_variables)]
```
np.ones instead of list comprehension

jbrockmendel replied:
+1.  also type above hanve-->have

josef-pkt replied:
see below, we want a list

jbrockmendel replied:
ok.  still have the typo

josef-pkt commented:
open naming issue: this still uses mostly `x` for the explanatory variable for which we estimate smooth terms, I use `exog_smooth`  for the created design matrix, i.e. smooth bases array.

I need a name to distinguish the original explanatory variable from it's transformation into the smooth basis representation.
I have no idea yet about a good name.
(more generally, I have an open issue suggesting extra `exog_transform` for other transformations that splines.)
I could also rename `exog_smooth` to something else and use it for the original exog.
???

(no change for now)

jbrockmendel commented:
Is this at a point where review would be useful?

josef-pkt commented:
> Is this at a point where review would be useful?

Yes, but it will still need another round later, after merging this PR.
E.g. pep-8 testing on travis currently fails. I'd like to fix the formatting problems that are left, but NOT the "unused variable" or "unused imports" cases.  Those are mostly things that need improvements or more unit tests, i.e. most of the unused variables or unused imports are supposed to be used, but I didn't get around to it.

Some improvements and refactorings will be follow-up issues, because I don't have the time and energy right now to review and finish every detail of code in this PR right now. (i.e. I am still unhappy about some options and implementation details, but those shouldn't prevent merging a first version of this.)

jbrockmendel commented:
> Yes, but it will still need another round later, after merging this PR.

Great, I'll take a look.

> but NOT the "unused variable" or "unused imports" cases. Those are mostly things that need improvements or more unit tests, i.e. most of the unused variables or unused imports are supposed to be used, but I didn't get around to it.

Since there is a decent chance something else will come up before you get around to it, please consider commenting these up with TODOs sufficiently for a non-you reader to figure out what is intended.

Unused imports obscure the dependency structure of the code.  Removing them is essentially costless.

jbrockmendel commented on statsmodels/base/_penalized.py:
```diff
@@ -54,13 +54,33 @@ def __init__(self, *args, **kwds):
+    def _handle_scale(self, params, **kwds):
+
+        #if kwds['scale'] is not None:
+        #    raise
```
raise something in particular?  why not just put `scale=None` in the function definition?

josef-pkt replied:
plain `raise` are most likely leftovers from debugging.
It's a simple way to create a break point for `pytest --pdb`, and sometimes I use it in an if-else to make sure I'm on the correct branch.

josef-pkt replied:
I changed it to `scale=None` keyword argument

jbrockmendel commented on statsmodels/base/_penalized.py:
```diff
@@ -54,13 +54,33 @@ def __init__(self, *args, **kwds):
+                mu = self.predict(params)
+                scale = self.estimate_scale(mu)
+            else:
+                #raise
```
delete

jbrockmendel commented on statsmodels/base/_penalties.py:
```diff
@@ -126,7 +126,7 @@ def func(self, params):
         return np.zeros(params.shape)
 
-    def hessian(self, params):
+    def deriv2(self, params):
```
gentle ping on this

jbrockmendel replied:
(also if the changes in this file are logically independent of the rest of the PR, doing them separately will make things easier on semi-literate reviewers like me)

josef-pkt replied:
I thought about extracting the bugfixes and I kept them in separate commits.
However, it's extra work, and in this PR the fixes are in the context where they are needed and indirectly tested by the usage in here. I didn't write specific unit tests for those fixes

jbrockmendel commented on statsmodels/base/tests/test_penalized.py:
```diff
@@ -714,8 +714,11 @@ def test_params_table(self):
         # failure for penalized values
         # check only first 2, others fails, see #4669
         exog_index = slice(None, 2, None)
+        exog_index = slice(None, None, None)
```
what should a future-reader make of the two versions of `exog_index` here?

josef-pkt replied:
good question,
the first version of exog_index and comments might now be obsolete.
My guess is that I fixed the underlying problem but had forgotten to clean up the unit test

josef-pkt replied:
I deleted the old part and extended to comment for the test method.
(When I wrote the test initially I thought TheilGLS sandwich is HC, but it's not)

```
        # override inherited because match is not good except for params and predict
        # The cov_type in GLMPenalized and in TheilGLS are not the same
        # both use sandwiches but TheilGLS sandwich is not HC
        # relative difference in bse up to 7%
```

josef-pkt replied:
`exog_index = slice(None, None, None)` makes indexing redundant but I keep it for consistency with the super method

jbrockmendel commented on statsmodels/compat/python.py:
```diff
@@ -232,3 +232,16 @@ def get_class(func):
+            return meta(name, bases, d)
+    return type.__new__(metaclass, 'temporary_class', (), {})
+
+from itertools import combinations
```
any reason not to put this at the top of the file with other imports?

josef-pkt replied:
Based on a file search, it looks like this is not used anymore.
I delete it and see what appveyor and travis say.

jbrockmendel replied:
I've got a PR that explicitly goes through compat and identifies unneeded stuff

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+from scipy import optimize
+from statsmodels.discrete.discrete_model import Logit
+from statsmodels.genmod.generalized_linear_model import (GLM, GLMResults,
+    GLMResultsWrapper, lm, _check_convergence)
```
`lm` is an alias for linear_model right?  maybe import that directly?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+    This inherits from GLMResults.
+
+    Warning: not all inherited methods might take correctly account of the
+    penalization
```
the phrasing of this sentence is awkard.  maybe "some inherited methods might not correctly take account of the penalization"?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+        self.normalized_cov_params = normalized_cov_params
+        self.scale = scale
+        edf = self.edf.sum()
+        self.df_model = edf - 1  # assume constant
```
why assume constant?  not a big deal here, but across the code I have trouble telling the driving logic behind when this is and is not used.

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+        self.df_model = edf - 1  # assume constant
+        # need to use nobs or wnobs attribute
+        self.df_resid = self.df_resid = self.model.endog.shape[0] - edf
+
```
`self.df_resid = ` is here twice.

Personally I find this clearer as `df_resid = nobs - (df_model + k_const)` where `k_const` is an indicator for the `# assume constant` comment.  Very nearly all the `df_resid`s in the codebase follow this pattern.

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+        # we are setting the model df for the case when super is using it
+        # df in model will be incorrect state when alpha/pen_weight changes
+        self.model.df_model = self.df_model
+        self.model.df_resid = self.df_resid
```
There was an issue/PR a while back about avoiding this pattern (setting model attributes from within results).  Are there no viable alternatives here?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+        transform : bool, optional
+            If transform is True, then the basis representation of the smooth
+            term will be constructed from the provided ``x``.
+        kwargs :
```
kwargs/kwds mismatch

jbrockmendel replied:
ditto below

jbrockmendel replied:
Not sure about the convention; should these have a `**` in the docstring?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+
+        Returns
+        -------
+        fig : matplotlib figure instance
```
capitalize Figure?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+        for iteration in range(maxiter):
+
+            # TODO: is this equivalent to point 1 of page 136:
+            # w = 1 / (V(mu) * g'(mu))  ?
```
This comment is very clear.  Good work.

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+            lin_pred += self._offset_exposure
+            mu = self.family.fitted(lin_pred)
+
+            # self.scale = self.estimate_scale(mu)
```
?

jbrockmendel replied:
can things of this nature be clarified?  it sounds like TODO isn't appropriate, but it also isn't something the reader is going to find useful for the foreseeable future.

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+                raise PerfectSeparationError(msg)
+
+            # TODO need atol, rtol
+            converged = _check_convergence(criterion, iteration, tol, 0)
```
pass kwargname=0 here?  Since my knowledge of _check_convergence is "it has an informative name so I can guess what it does, but I don't know its signature off the top of my head", this is the only thing unclear about this line.

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+            if converged:
+                break
+        self.mu = mu
+        self.scale = self.estimate_scale(mu)
```
as before, setting mu and scale strictly necessary?  Easier to avoid now than revisit years down the road.

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+                                    cov_type=cov_type, cov_kwds=cov_kwds,
+                                    use_t=use_t)
+
+        glm_results.method = "PIRLS"
```
Can any of these be set in `GLMGAMResults.__init__`?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+
+        Warning: In many case the optimization might converge to local
+        optima or near optima. Different start_params or using a global
+        optimizer is recommendet, default is basinhopping.
```
"recommended".  should optima be singular here?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+        criterion='aic'
+            name of results attribute to be minimized.
+            Default is 'aic', other options are 'gcv', 'cv' or 'bic'.
+
```
reason for this extra newline?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+        see also https://github.com/statsmodels/statsmodels/issues/5381.
+
+        This does currently not use any analytical derivatives for the
+        criterion minimization.
```
"does not currently"?

jbrockmendel commented on statsmodels/gam/gam.py:
```diff
@@ -0,0 +1,767 @@
+    aug_y, aug_x, aug_weights = make_augmented_matrix(y, x, 2 * s, weights)
+    wls_results = lm.WLS(aug_y, aug_x, aug_weights).fit()
+    # MinimalWLS does not return normalized_cov_params
+    # wls_results = reg_tools._MinimalWLS(aug_y, aug_x, aug_weights).fit()
```
this comment could use clarification.  is it saying "we would use MinimalWLS, but [...] 
 If we did, this would be how we call it [...]"?

jbrockmendel replied:
nice, thanks

jbrockmendel commented on statsmodels/gam/gam_cross_validation/cross_validators.py:
```diff
@@ -0,0 +1,73 @@
+            number of folds
+
+        Examples
+        --------
```
fill in or delete

jbrockmendel commented on statsmodels/gam/gam_cross_validation/cross_validators.py:
```diff
@@ -0,0 +1,73 @@
+        All the folds have size trunc(n/k), the last one has the complementary
+        """
+
+        self.nobs = None
```
is this going to be used?

jbrockmendel replied:
Looks like nobs is defined as X.shape[0] below; this may be unnecessary

jbrockmendel commented on statsmodels/gam/gam_cross_validation/gam_cross_validation.py:
```diff
@@ -0,0 +1,170 @@
+
+class MultivariateGAMCV(BaseCV):
+    def __init__(self, smoothers, alphas, gam, cost, y, cv):
+        # the gam class has already an instance
```
can you flesh this comment out?  im confused in several directions

jbrockmendel replied:
"TODO: figure out what the end of the next sentence is supposed to be"?  if it isn't clear to you, its definitely not clear to the reader

josef-pkt replied:
I delete the comment as misleading here

jbrockmendel commented on statsmodels/gam/gam_cross_validation/gam_cross_validation.py:
```diff
@@ -0,0 +1,170 @@
+        self.cv_std = None
+
+    def plot_path(self):
+        if have_matplotlib:
```
otherwise raise?

jbrockmendel replied:
gentle ping

josef-pkt replied:
class is a stub and currently unused

jbrockmendel commented on statsmodels/gam/gam_cross_validation/gam_cross_validation.py:
```diff
@@ -0,0 +1,170 @@
+            plt.plot(self.alphas, self.cv_error, 'o', c='black')
+            plt.plot(self.alphas, self.cv_error + 1.96 * self.cv_std, 'o',
+                     c='blue')
+            plt.plot(self.alphas, self.cv_error - 1.96 * self.cv_std, 'o',
```
is cv_error \pm 1.96 * self.cv_std worth calculating just once at the top?

jbrockmendel commented on statsmodels/gam/gam_penalties.py:
```diff
@@ -0,0 +1,309 @@
+    """
+
+    def __init__(self, univariate_smoother, alpha=1, weights=1):
+        self.weights = weights  # should we keep weights????
```
"TODO"

jbrockmendel commented on statsmodels/gam/gam_penalties.py:
```diff
@@ -0,0 +1,309 @@
+            raise ValueError('all the input values should be list of the same'
+                             ' length. len(smoothers)=',
+                             len(multivariate_smoother.smoothers),
+                             ' len(alphas)=', len(alpha))
```
use string formatting?

jbrockmendel commented on statsmodels/gam/gam_penalties.py:
```diff
@@ -0,0 +1,309 @@
+            self.weights = weights
+
+        self.mask = [np.array([False] * self.k_params)
+                     for _ in range(self.k_variables)]
```
np.zeros, also could just make this a single 2D array?

josef-pkt replied:
the top level of those are the different spline components, we keep those as a list (currently). They are always used in loops, so list is more efficient, than arrays, and avoids the issue of column versus row orientation.

josef-pkt replied:
change to list of `zeros`

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+"""
+
+from __future__ import division
+# import useful only for development
```
which imports is this referring to?

jbrockmendel replied:
gentle ping

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+from statsmodels.tools.linalg import matrix_sqrt, transf_constraints
+
+
+# Obtain b splines from patsy
```
does this comment refer to _equally_spaced_knots?

jbrockmendel replied:
if so, put it in docstring?  or at least immediately-adjacent to the relevant function?

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+    return quantiles.reshape(probs.shape, order="C")
+
+
+# from patsy splines.py
```
is this modified from the original?  if not, why not just import?

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+
+
+def make_bsplines_basis(x, df, degree):
+    ''' make a spline basis for x '''
```
triple-double

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+    return covd2
+
+
+# TODO: this function should be deleted
```
planning on actually doing this?

jbrockmendel replied:
gentle ping

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+# result = np.array(result)
+# print(result.shape)
+# plt.plot(result.T)
+# plt.show()
```
does this belong in an example or test or something?

jbrockmendel replied:
gentle ping

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+        if constraints == 'center':
+            constraints = base4[0].mean(0)[None, :]
+
+        if constraints is not None and not isinstance(constraints, str):
```
string_types?

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+        basis, der_basis, der2_basis = _eval_bspline_basis(
+            self.x, self.knots, self.degree,
+            include_intercept=self.include_intercept)
+        # cov_der2 = np.dot(der2_basis.T, der2_basis)
```
document or delete

jbrockmendel replied:
gentle ping

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+        else:
+            ctransf = np.eye(basis.shape[1])
+        # use np.eye to avoid rescaling
+        # ctransf = np.eye(basis.shape[1])
```
still needed?

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+        else:
+            self.transf_mean = np.zeros(basis.shape[1])
+        s = self._splines_s()[:-1, :-1]
+        if not self.constraints == 'none':
```
`not foo == bar` --> `foo != bar`, ditto a few lines up

jbrockmendel replied:
gentle ping

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+    def _smooth_basis_for_single_variable(self):
+        basis = dmatrix("cc(x, df=" + str(self.df) + ") - 1", {"x": self.x})
+        self.design_info = basis.design_info
+        n_inner_knots = self.df - 2 + 1  # +n_constraints
```
what should the reader make of `# +n_constraints`?

jbrockmendel replied:
gentle ping... gonna stop pointing these all out now

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+        super(BSplines, self).__init__(x, include_intercept=include_intercept,
+                                       variable_names=variable_names)
+
+
```
extra newline

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+
+        return smoothers
+
+# class CubicRegressionSplines(BaseCubicSplines):
```
sandbox?

jbrockmendel commented on statsmodels/gam/tests/results/results_mpg_bs.py:
```diff
@@ -0,0 +1,1281 @@
+
```
document where these results came from?

jbrockmendel commented on statsmodels/gam/tests/results/results_mpg_bs_poisson.r:
```diff
@@ -0,0 +1,49 @@
+
+source("M:\\josef_new\\eclipse_ws\\statsmodels\\statsmodels_py34_pr\\tools\\R2nparray\\R\\R2nparray.R")
```
is there a viable way to make this more general?

jbrockmendel commented on tools/R2nparray/R/R2nparray.R:
```diff
@@ -63,7 +63,7 @@ convert.data.frame <- function(X, name='default', strict=TRUE){
     if (length(X) > 1){
         head = paste(name, '= np.array([\n')
-        mid = strwrap(paste(X, collapse=', '), width=76, prefix='    ')
+        mid = strwrap(paste(X, collapse=', '), width=79, indent=3, exdent=4)
```
As long as you're tinkering in this file, on line 192 there is an extra space before an "=" that would be soo nice to get rid of.

jbrockmendel commented on statsmodels/tools/linalg.py:
```diff
@@ -75,3 +75,101 @@ def stationary_solve(r, b):
+    """
+
+    from scipy import linalg
+    # constraints = spsc._constraints
```
?

josef-pkt replied:
deleted, I guess left over during copy-paste from example

jbrockmendel commented on statsmodels/stats/contrast.py:
```diff
@@ -156,11 +156,13 @@ def summary(self, xname=None, alpha=0.05, title=None):
-                   (repr(self.fvalue), self.pvalue, self.df_denom, self.df_num)
+            return '<F test: F=%s, p=%s, df_denom=%.3g, df_num=%.3g>' % \
+                   (repr(self.fvalue), self.pvalue, self.df_denom,
+                    self.df_num)
```
prettify by putting in parens and removing backslash?

josef-pkt replied:
changed, given that I had to edit these lines already

jbrockmendel commented on statsmodels/gam/tests/test_penalized.py:
```diff
@@ -0,0 +1,571 @@
+    def test_fitted(self):
+        res1 = self.res1
+        pred = res1.get_prediction()
+        self.rtol_fitted = 1e-7
```
is there a reason to make this an attribute?

josef-pkt replied:
no special reason, most likely to continue the pattern that inherit from a `Check` class.
I will leave this as is.

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+
+
+def test_partial_plot():
+    # TODO: SMOKE No test is performed.
```
`@pytest.mark.smoke`?

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+
+    assert_allclose(hat_y, data_from_r["y_est"], rtol=0, atol=0.008)
+    # TODO: bug missing scale
+    bug_fact = np.sqrt(res_glm_gam.scale) * 0.976  # 0.106
```
what should a reader making of this 0.106?

josef-pkt replied:
bug_fact = 0.1059984985594
I edit the comment `# this is = 0.106`

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+                              disp=0, maxiter=5000)
+    glm = GLM(y, bsplines.basis)
+
+    # case with constant column in exog is currently wrong
```
TODO?

josef-pkt replied:
I think this is outdated, needs verification

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    y = x[:, 0] * x[:, 0] + np.random.normal(0, .01, n)
+    y -= y.mean()
+    alpha = 0.0
+    # BUG: mask is incorrect if exog is not None, start_idx missing
```
is "BUG" used frequently?  I commonly grep for "TODO" or "FIXME" when looking for things that need attention.

josef-pkt replied:
this needs another look and verifying what the current code does.
Right now I'm just looking at easy comments/changes.

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    assert_allclose(spl_s_R, spl_s, atol=4.e-10)
+
+
+def test_partial_values2():
```
what differentiated this case from test_partial_values1?  i.e. can this be named `test_partial_values_interesting_case`?

josef-pkt replied:
When there are several test cases without specific focus, then I just number them.

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    # plt.plot(y_est_pirls)
+    # plt.plot(y_est_glm)
+    # plt.plot(y, '.')
+    # plt.show()
```
below there is a nice comment on similarly commented-out code explaining why its there

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    # TODO: if alpha changes in pirls this should be updated
+
+    gam = GLMGam(y, smoother=ccs, alpha=alpha)
+    # gam_res = gam._fit_pirls(y, ccs, alpha=alpha)
```
?

josef-pkt replied:
old version, deleted
signature for _fit_pirls has also changed

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    alpha = 1
+    aug_y, aug_x, aug_w = make_augmented_matrix(y, x, s, w)
+    rs = matrix_sqrt(alpha * s)
+    # rs = sp.linalg.cholesky(alpha * s)
```
?

josef-pkt replied:
I leave it in with an extra comment

jbrockmendel replied:
looks good, thanks,

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    degree = [5]
+    bsplines = BSplines(x, degree=degree, df=df)
+    # y_mgcv is obtained from R with the following code
+    # g = gam(y~s(x, k = 10, bs = "cr"), data = data, scale = 80)
```
this comment is great.  positive reinforcement!

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+
+
+def test_multivariate_gam_cv():
+    # SMOKE test
```
pytest.mark.smoke?

josef-pkt replied:
changed

josef-pkt replied:
I dropped this again. pytest.mark.smoke is not yet available

jbrockmendel replied:
all it takes is one line in setup.cfg

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    glm_gam = GLMGam(y, exog=np.ones((len(y), 1)), smoother=bsplines,
+                     alpha=alpha)
+#     res_glm_gam = glm_gam.fit(method='nm', max_start_irls=0,
+#                               disp=1, maxiter=10000, maxfun=5000)
```
?

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    alpha = 0.00002
+    # gp = UnivariateGamPenalty(alpha=alpha, univariate_smoother=bsplines)
+    # lg_gam = LogitGam(y, bsplines.basis, penal=gp)
+    #
```
?

josef-pkt replied:
TODO: this needs another look comparing LogitGAM and GLMGAM

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    # plt.plot(x, y_mgcv, '.', label='mgcv')
+    # plt.plot(x, y, '.', label='y')
+    # plt.legend()
+    # plt.show()
```
if deleting these commented-out bits isn't an option, might as least make them into a single function instead of copying them all over the place

josef-pkt replied:
another time, 
needs checking whether we want to delete it or not, which I haven't done.
deletion of the commented out plots is likely

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+    # alpha = 1000
+    alpha = 0.03
+    alpha = 0.0001
+    alpha = 0.1
```
maybe just one of these 4?

jbrockmendel commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,791 @@
+from statsmodels.gam.gam_cross_validation.cross_validators import KFold
+from statsmodels.genmod.generalized_linear_model import GLM
+from statsmodels.genmod.families.family import Gaussian
+from statsmodels.genmod.generalized_linear_model import lm
```
import directly from statsmodels.regression

josef-pkt commented on statsmodels/genmod/generalized_linear_model.py:
```diff
@@ -792,7 +796,7 @@ def estimate_scale(self, mu):
             elif self.scaletype.lower() == 'dev':
                 return (self.family.deviance(self.endog, mu, self.var_weights,
-                                             self.freq_weights, self.scale) /
+                                             self.freq_weights, 1.) /
```
Yes, this fixes the related bugs/problems, unit tests should cover this because it was needed for the GAM extension.

josef-pkt commented:
@jbrockmendel  Thanks for the review.
I added some general replies to some of the review comments.
For the remaining review comments I have to go through the code and make the relevant changes or reply directly.
I hope to go back to working on this PR soon so it can be merged.

josef-pkt commented on statsmodels/genmod/tests/test_glm.py:
```diff
@@ -1114,7 +1138,14 @@ def test_gradient_irls():
                     ehess = mod_gradient.hessian(rslt_gradient.params, observed=False)
                     gradient_bse = np.sqrt(-np.diag(np.linalg.inv(ehess)))
                     assert_allclose(gradient_bse, rslt_irls.bse, rtol=1e-6, atol=5e-5)
+                    # this uses observed=True
```
AFAIU after a month, "this" refers to `rslt_irls.bse`  (based on normalized_cov_params in WLS)
I edited the comment

josef-pkt commented on statsmodels/genmod/tests/test_glm.py:
```diff
@@ -282,6 +284,28 @@ def test_compare_OLS(self):
+        cls.decimal_bic = DECIMAL_0
+        cls.decimal_bse = DECIMAL_2
+
+        from statsmodels.datasets.longley import load
```
I don't want to break the pattern for the `test_glm` module with loading the examples in the test classes instead of as module globals
glm and discrete cover many different cases and I never checked whether loading the datasets and results only once would be worth the extra memory usage.

josef-pkt commented on statsmodels/gam/tests/test_penalized.py:
```diff
@@ -0,0 +1,571 @@
+
+        cls.rtol_fitted = 1e-12
+        # cls.covp_corrfact = 1.0025464444310588  # without edf
+        # edf is implemented
```
historical, but I want to keep it as reference
corr_fact is the difference to R if we use the standard df_resid (1.0025 means the difference is small in this case)
but now we have edf and it is not needed anymore

jbrockmendel replied:
ok, but that needs to be made clear to the reader

josef-pkt commented on statsmodels/gam/tests/test_penalized.py:
```diff
@@ -0,0 +1,571 @@
+        pred = res1.get_prediction()
+        self.rtol_fitted = 1e-7
+        pls6_fittedvalues = np.array([
+            2.45008146537851, 3.14145063965465, 5.24130119353225,
```
All the results are from R mgcv, unless regression test is explicitly mentioned.
These are cases, where I just wanted to test a few numbers from an interactive R session instead of putting all those numbers in csv files.
I'm not sure whether I have a log of the corresponding R commands.

josef-pkt commented:
I'm not sure what to do about the `test_partial_values` and `test_partial_plot`
I didn't rewrite them from the original version, just got them to pass. Those unit tests are not very good because we don't replicate mgcv closely, e.g. I don't know what are the penalization parameter and knots that corresponds to the mgcv results.

However, it looks like I didn't add any unit tests for `partial` in my new versions. But AFAIR I verified them in my notebook examples.

josef-pkt commented:
BUG/ENH: `GLMGAMResults.plot_partial` uses hardcoded confidence level alpha, i.e. `1.96 * se`

josef-pkt commented:
I forgot to protect plot test for availability of matplotlib
amended and force pushed

(master has a failing tsa test on one travis)

josef-pkt commented on statsmodels/gam/tests/test_gam.py:
```diff
@@ -0,0 +1,808 @@
+
+    glm_gam = GLMGam(y, smoother=bsplines, alpha=alpha)
+    res_glm_gam = glm_gam.fit(method='bfgs', max_start_irls=0,
+                              disp=1, maxiter=10000, maxfun=5000)
```
looks like a duplicate 
maybe I had needed it for start_params with "nm", bfgs works fine here

jbrockmendel commented:
@josef-pkt lots of comments coming in, most of which I assume are more notes to yourself than to me.  Let me know when this is ready for another review and I'll read through comments en masse then.

josef-pkt commented:
I amended and force pushed after adding more cleanup changes in response to the review comments.

I'm about ready to stop with this, but there are a few open questions, e.g. things where I don't remember what the status is (e.g. comparison GLMGAM, GLMPenalized, LogitGAM). And I have not looked carefully at the k-fold cross-validation.

josef-pkt commented:
@jbrockmendel  I'm not making any more changes for now, so you can have a look without having a moving target.

I will need to go back to some of my notebook examples to check the behavior for some cases and parts.

jbrockmendel commented on statsmodels/base/_penalized.py:
```diff
@@ -54,13 +54,28 @@ def __init__(self, *args, **kwds):
+                scale = self.estimate_scale(mu)
+            else:
+                scale = 1
+                pass
```
delete

jbrockmendel commented on statsmodels/base/_penalized.py:
```diff
@@ -54,13 +54,28 @@ def __init__(self, *args, **kwds):
+
+        return scale
+
+
```
1 line too many

jbrockmendel commented on statsmodels/base/_penalized.py:
```diff
@@ -98,18 +114,20 @@ def score(self, params, pen_weight=None, **kwds):
             pen_weight = self.pen_weight
 
-        sc = super(PenalizedMixin, self).score_obs(params, **kwargs)
+        sc = super(PenalizedMixin, self).score_obs(params, **kwds)
```
why this change?

jbrockmendel replied:
OK.  If this could be made consistent across the code that'd be swell, but clearly not necessary for this PR.

jbrockmendel commented on statsmodels/base/model.py:
```diff
@@ -1686,12 +1686,16 @@ def wald_test(self, r_matrix, cov_p=None, scale=1.0, invcov=None,
 
+        # TODO streamline computation, we don't need to compute J if given
+        if df_constraints is not None:
+            # let caller override df_constraint. J
```
is the trailing `J` here a typo?

jbrockmendel commented on statsmodels/base/tests/test_penalized.py:
```diff
@@ -705,17 +705,20 @@ def _initialize(cls):
         # override inherited because match is not good except for params and predict
+        # The cov_type in GLMPenalized and in TheilGLS are not the same
+        # both use sandwiches but TheilGLS sandwich is not HC
+        # relative difference in bse up to 7%
```
in places where there is potential ambiguity about what is a multi-line comment vs separate comments, I've started using the pattern:

```
# some comment [...]
#  continuation of the comment gets an extra space
# Separate comment no extra space
```

jbrockmendel commented on statsmodels/gam/gam_cross_validation/cross_validators.py:
```diff
@@ -0,0 +1,73 @@
+    """
+    def __init__(self):
+
+        return
```
this seems unnecessary

jbrockmendel commented on statsmodels/gam/gam_cross_validation/cross_validators.py:
```diff
@@ -0,0 +1,73 @@
+
+        folds = np.array_split(index, self.k_folds)
+        for fold in folds:
+            test_index = np.array([False]*nobs)
```
np.zeros(nobs, dtype=np.bool_)

jbrockmendel commented on statsmodels/iolib/summary.py:
```diff
@@ -22,6 +22,16 @@ def forg(x, prec=3):
                          .format(prec=prec))
 
 
+def d_or_f(x, width=6):
```
docstring?

josef-pkt replied:
added

jbrockmendel commented on tools/R2nparray/R/R2nparray.R:
```diff
@@ -186,10 +186,12 @@ mkhtest <- function(ht, name, distr="f") {
     def __init__(self, **kw):
         dict.__init__(self, kw)
-        self.__dict__  = self
+        self.__dict__ = self
```
This makes my day, thank you.

josef-pkt commented:
capitalization in class names?
 we currently have `GLMGam` but `GLMGAMResults`

I guess it would be better to switch to `GLMGAM`, and corresponding LogitGAM, PoissonGAM and similar when we have them. WLSGAM ?

In favor of `Gam` is that we read it as a word and not as a sequence of letters, e.g. word Anova versus acronym ANOVA. (And I don't like to use the shift key for capital letters in sequence.)
`GLMGamResults` looks prettier (pronounced "ge el em gam results" :)

josef-pkt commented:
@david-waterworth Thanks for the feedback. (It's good to not have to make all decisions by myself)

I changed to GLMGamResults.

Another consideration:
We use `GLM` for the specific model class based on linear exponential family.
The plan is to use `Gam` in a more general sense of models with a linear predictor that has additive nonparametric components, i.e. we will have `XXXGam` that don't have an equivalent GLM family.

GLM in GLMGAM would be redundant or tautological in statistics, i.e. GAM is a specific case of penalized GLM. However, we use a specific definition of "generalized" in GLM, but a generic "generalized" in "Gam".

josef-pkt commented:
I pushed the first set of changes to trigger CI testing. But I will continue tomorrow, (most likely with amending the last commit)..

david-waterworth commented:
> GLM in GLMGAM would be redundant or tautological in statistics, i.e. GAM is a specific case of penalized GLM. However, we use a specific definition of "generalized" in GLM, but a generic "generalized" in "Gam".

That is consistent with R libraries say mboost which have separate gamboost and glmboost classes.

josef-pkt commented on statsmodels/iolib/summary.py:
```diff
@@ -312,11 +322,11 @@ def num_to_str(x, width=6):
+          ('Df Model:', lambda: [d_or_f(results.df_model)]),
           #TODO: check when we have non-integer df
-          ('Df Residuals:', lambda: [num_to_str(results.df_resid)]),
+          ('Df Residuals:', lambda: [d_or_f(results.df_resid)]),
```
I cleaned up the comments in this neighborhood (not entire module)

josef-pkt commented:
amended and force pushed the second part
I think I addressed all points except for the cross-validator, `gam_cross_validation.py`. I don't want to deviate there from the original version for now because I didn't try to figure out the details how that part works.

I'm planning to merge this soon.

jbrockmendel commented on statsmodels/gam/gam_cross_validation/gam_cross_validation.py:
```diff
@@ -0,0 +1,170 @@
+        self.y = y
+        self.train_test_cv_indices = self.cv.split(self.x, self.y, label=None)
+
+    def fit(self, **kwargs):
```
kwds?

jbrockmendel commented on statsmodels/gam/gam_cross_validation/gam_cross_validation.py:
```diff
@@ -0,0 +1,170 @@
+        self.cv_std = np.zeros(shape=(len(self.alphas_grid, )))
+        self.alpha_cv = None
+
+    def fit(self, **kwargs):
```
kwds?

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+
+
+def _R_compat_quantile(x, probs):
+    # return np.percentile(x, 100 * np.asarray(probs))
```
"equivalent to returning np.percentile... but we don't do that because of ..."

jbrockmendel commented on statsmodels/gam/smooth_basis.py:
```diff
@@ -0,0 +1,821 @@
+    if deriv == 'all':
+        return basis, der1_basis, der2_basis
+    else:
+        return ret
```
can we avoid this multiple-return?   these just cause hassle down the road

jbrockmendel commented:
> I'm planning to merge this soon.

The remaining unclear-comments (most but not all of which I just gentle-pinged on) are strictly harder to resolve in follow-ups.  Everything else I can kid myself that follow-ups will be allowed to improve.

josef-pkt commented:
inheritance is nice
The inherited `from_formula` for exog with spline components in keyword works without changes.
The results are the same as without formula, but we don't get the correct exog_names.

(I thought I had changed the code to create nicer names for the spline components, but I don't find any related committed code.)

```
gam_ccf = GLMGam.from_formula('city_mpg ~ fuel + drive', data=df_autos2, smoother=cc, alpha=(1 / s_scale * sp))
print(gam_ccf.exog_names)
['const', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9', 'x10', 'x11', 'x12']
```

josef-pkt commented:
using pandas DataFrame for splines still does not work

```
x_spline = df_autos2[['weight', 'hp']]
cc = CyclicCubicSplines(x_spline, df=[6, 5], constraints='center')
```
this raises an exception in indexing.
using numpy array `x_spline.values` works and was used in my initial examples.

using only one spline component also breaks, but somewhere else (assignment to shape to make it 2-D)
```
x_spline = df_autos2['hp']
cc = CyclicCubicSplines(x_spline, df=[6], constraints='center')
cc.variable_names
```

i.e. we need an `asarray` somewhere in the spline creation.

josef-pkt commented:
problem with patsy and pandas integration in GLMGAM

`GLMGAM.__init__` uses asarray(exog) and throws away any pandas names and design_info

```
        if exog is not None:
            exog_linear = np.asarray(exog)
            k_exog_linear = exog_linear.shape[1]
```

We are not going through the super data handling because we have to add extra terms to the `exog`

- proper solution: create dataframe and design_info for the extended, full exog. converting to pandas would be ok, creating a full design_info is difficult.

- partial solution: just create exog_names before throwing away the information with asarray. We can also store the design_info for `exog_linear` for future use. The problem is that we will need to duplicate some of the code for data handling in the GAM classes.

josef-pkt commented:
latest commit adds handling of formula and pandas exog names, so that we get the following in a GLMGAM model created with from_formula

```
res_ccf.model.exog_names
['Intercept',
 'fuel[T.gas]',
 'drive[T.fwd]',
 'drive[T.rwd]',
 'weight_s0',
 'weight_s1',
 'weight_s2',
 'weight_s3',
 'weight_s4',
 'hp_s0',
 'hp_s1',
 'hp_s2',
 'hp_s3']
```

This works but it is not a full, clean or generic solution yet, see #5469

josef-pkt commented:
another messy piece

base.model.Model.from_formula attaches `formula` to model after creating the model
`mod.formula = formula`

This mean I cannot delete it in a `model.__init__`, and I need to delete it in `fit`.
I need to remove the formula attribute because it messes up the predict transform, because `formula` is only for the linear component and we don't have an underlying model.data.design_info for the full model.

I can delete the `design_info` in `__init__`. It is also attached, but through `__init__` kwds.
design_info needs to be removed because it is incorrect for `wald_test_terms`. I don't have unit tests for the other wald tests.

I don't have yet transform handling by patsy for the linear exog in predict. `transform` keyword applies to the spline terms, ie. creating the basis function columns for new spline exog.  `exog_smooth` (?)
I guess I will have to copy the transform part from base predict to gam predict.

The only alternative would be to have a data instance for the linear part with full support of additional methods, but that still requires that we use the transform in predict separately for the different parts and super/base predict can only handle the default design_info and transform.

josef-pkt commented:
latest commit `ENH: add predict transform for exog_linear, more pandas returns`
copies and adjusts most of the code from base.model.Model.predict to a standalone function with additional keywords.
This is currently in gam.py and I did not try to DRY out this code duplication.

predict now applies separate transforms to exog_linear and exog_smooth
(renamed `x` -> `exog_smooth` in predict and get_prediction.

I added the wrapper for the results class, but it does not add pandas indices yet because the created full exog is just numpy.
To return pandas wrapped arrays, we need a data.PandasData instance.

The next commit will add conversion to pandas DataFrame for the full exog before calling super, that makes PandasData available and the wrapper uses pandas.

It's likely that this data and formula handling can be restructured and streamlined, and we will get other cases with similar pattern.
I didn't do a full "design review".  
TDD: add unit tests and make the code work

josef-pkt commented:
python 2.7 is picky in not allowing reusing the wrapper across classes, no problems in python 3

e.g. https://travis-ci.org/statsmodels/statsmodels/jobs/481501429


```
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        results = object.__getattribute__(self, '_results')
        data = results.model.data
        if how and isinstance(how, tuple):
            obj = data.wrap_output(func(results, *args, **kwargs), how[0], how[1:])
        elif how:
>           obj = data.wrap_output(func(results, *args, **kwargs), how)
E           TypeError: unbound method cov_params() must be called with GLMGamResults instance as first argument (got GLMResults instance instead)
```

in `gam.py` I have
`wrap.populate_wrapper(GLMResultsWrapper, GLMGamResults)`

the old code creates a duplicating subclass, e.g. in discrete we have

```
class CountResultsWrapper(lm.RegressionResultsWrapper):
    pass
wrap.populate_wrapper(CountResultsWrapper, CountResults)
```

I open a new issue for this

josef-pkt commented:
BUG (?), I guess this is not intended
`fit(method='irls')`` is not ruled out. It looks like it delegates to the inherited `_fit_irls` and returns a plain GLMResults and not a GLMGamResults

```
r = glm_gam.fit(method='irls', max_start_irls=0, disp=1, maxiter=10000, maxfun=5000)
type(r), type(r._results)

(statsmodels.genmod.generalized_linear_model.GLMResultsWrapper,
 statsmodels.genmod.generalized_linear_model.GLMResults)
```

based on test case in test_gam `test_multivariate_gam_cv_path` which uses irls and not pirls

rename fit_pirls to fit_irls to override the inherited method?

I tried a K-fold cross validation case from the test case and it looks not user friendly, i.e. too many extra steps that are not predefined.

josef-pkt commented:
next change:
- `fit(method='irls')` also uses pirls
- `max_start_irls` > 0 now uses fit_pirls instead of letting super fit handle this which would use the unpenalized `_fit_irls`. Comparing the run times for unit test before and after this change, this might increase the time a little bit, maybe by half a second for gam/tests which takes around 11.5 seconds.
We currently only have test cases where the unpenalized estimation overfits but doesn't have numerical problems. 

docstring for fit does not list all the arguments, it mainly says "most parameters are the same as for GLM"

josef-pkt commented:
`max_start_irls` seems to be ignored anyway, in super fit

`PenalizedMixin.fit` has
```
        if hasattr(self, 'family'):  # assume this identifies GLM
            kwds.update({'max_start_irls' : 0})
```

fit with gradient returns a GLMResultsWrapper
`GLM.fit`, `_fit_gradient` (super of GLMGam.fit with scipy optimizers uses the correct results class
```
        results_class = getattr(self, '_results_class', GLMResults)
        glm_results = results_class(self, rslt.params,
```
but uses hard coded wrapper
`        return GLMResultsWrapper(glm_results)`

with method='bfgs'

```
type(res_glm_gamgcv), type(res_glm_gamgcv._results)
(statsmodels.genmod.generalized_linear_model.GLMResultsWrapper,
 statsmodels.gam.gam.GLMGamResults)
```

in discrete count models we also have the wrapper class attached for generic use
e,g, ZeroInflatedPoisson attaches the classes in `__init__` and not as class attributes
```
        self.result_class = ZeroInflatedPoissonResults
        self.result_class_wrapper = ZeroInflatedPoissonResultsWrapper
        self.result_class_reg = L1ZeroInflatedPoissonResults
        self.result_class_reg_wrapper = L1ZeroInflatedPoissonResultsWrapper
```

I think or IIRC one advantage of defining it in `__init__` is that it is runtime and the classes don't have to be available when the class is created. gam.py module has results class before defining model class, most other modules have this reversed.

GLM uses a leading underscore for `_results_class`
That's not in master, so we can still change it in GLM

josef-pkt commented:
`gam_cross_validation.BasePenaltiesPathCV` is unused
I think it is supposed to be the base class for MultivariateGAMCVPath but it is not subclassed and has different `__init__` arguments. The only method of BasePenaltiesPathCV is a plot method.
I'm not sure how the plot method is supposed to work with more than 1 component, i.e. multi-dimensional alpha grid.

ignore for now.

MultivariateGAMCVPath works fine in a simple example (1 component and no exog_linear), but API should be improved.
`_split_train_test_smoothers` and call to create a model do not support exog_linear yet.

in general: Can we use a short cut and just split the existing model `exog`, i.e. including all components, instead of recreating smoothers?
The problem is that we need to use the signature of GLMGam, which requires smoothers.

my `GLMGam.select_penweight` is simpler because it doesn't resample the data.

josef-pkt commented on statsmodels/gam/gam_cross_validation/gam_cross_validation.py:
```diff
@@ -0,0 +1,170 @@
+            self.cv_std[i] = cv_err.std()
+
+        self.alpha_cv = self.alphas_grid[np.argmin(self.cv_error)]
+        return self
```
not necessary, but more convenient
fit always returns something, even though in this case the changes are inplace

josef-pkt commented:
naming again
GLMGam uses `smoother` instead of `smoothers` . It should be plural it's a list with one or more smoother instances.
(I just mis-typed it during refactoring.)

WRONG: GLMGam has one additive, multi-component smoother, which holds the list of individual smooth components.
However, alphas and similar are a list in GLMGam corresponding to the list of `smoothers` in the additive `smoother`.

josef-pkt commented:
folder names: too many identical names
```
from gam.gam import GLMGam
import gam.api as gam
from .gam_cross_validation.gam_cross_validation import MultivariateGAMCVPath
```

josef-pkt commented:
all red, I forgot to add the `api.py` module, 
amended and force pushed

josef-pkt commented:
being a bit pedantic:
smooth_basis classes are not "Smoothers" but it's in their name.
`nonparametric.smoother` are model classes that actually estimate a smooth function, i.e. `Smoothers` smooth.
Here the `XXXSmoother` classes just define smooth basis functions and supporting characteristics.

The main user classes don't have smooth in the name: `BSplines`, `CyclicCubicSplines`, but all univariate, single component classes do.

josef-pkt commented:
strange test failure in all CI that I don't see on my computer. It looks like exog_linear is penalized based on edf. But other parts don't fail.

I added a few asserts that select_penweight and select_penweight_kfold don't affect some of the penalization model attributes

```
____________________________ TestGAMMPGBS.test_edf _____________________________
self = <statsmodels.gam.tests.test_penalized.TestGAMMPGBS object at 0x7fd3c2852550>
    def test_edf(self):
        res1 = self.res1
        res2 = self.res2
>       assert_allclose(res1.edf, res2.edf_all, rtol=1e-6)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-06, atol=0
E       
E       (mismatch 100.0%)
E        x: array([ 0.977614,  0.977614,  0.977614,  0.977614,  0.699172,  0.924357,
E               0.913044,  0.890399,  0.897879,  0.915949,  0.945958,  0.956598,
E               0.957664,  0.95133 ,  0.971678,  0.201906,  0.131707,  0.127122,
E               0.209266,  0.419033,  0.469991,  0.712416,  0.454576,  0.838797])
E        y: array([ 1.      ,  1.      ,  1.      ,  1.      ,  0.715183,  0.945523,
E               0.933952,  0.910788,  0.91844 ,  0.936923,  0.96762 ,  0.978503,
E               0.979594,  0.973115,  0.993929,  0.20653 ,  0.134723,  0.130033,
E               0.214058,  0.428629,  0.480753,  0.72873 ,  0.464985,  0.858005])
statsmodels/gam/tests/test_penalized.py:468: AssertionError
```

josef-pkt commented:
there is still the same failure in at least one of the runs, but the new asserts don't fail
https://travis-ci.org/statsmodels/statsmodels/jobs/485453073

in other news: the latest pandas release broke the VAR unit tests

josef-pkt commented:
The discrepancy in the travis failure is proportional, so most likely a problem with `scale`

```
xx = np.array([ 0.977614,  0.977614,  0.977614,  0.977614,  0.699172,  0.924357])
yy = np.array([ 1.      ,  1.      ,  1.      ,  1.      ,  0.715183,  0.945523])
yy / xx
array([ 1.02289861,  1.02289861,  1.02289861,  1.02289861,  1.02289994,
        1.02289808])
```

I'm not able to reproduce or reverse engineer that number yet. Possibly something changes model.scale or model.scaletype.

Also, there is potentially a circular definition
- computing edf uses scale
- computing scale might use edf as df_resid

I'm not sure how this currently works. 
`GLMGamResults.__init__`
- attach scale given by args
- compute edf
- set df_model, df_resid based on edf, including the one in the model `self.model.df_resid = self.df_resid`
- recompute scale, `self.scale = scale = self.model.estimate_scale(mu)`

This sounds bad, but I think (!) there is a invariance in the edf so that it is independent of scale as long as we use the same scale in all computation.

However, the failure seems to depend on the random sequence of how the unit tests are run by pytest, so my guess is still some unintended changes somewhere.

josef-pkt commented:
Ok. 
`gam_bsf.select_penweight` changes `res_bsf.model.scale`, but not `res_bsf.scale` and `res_bsf.model.scaletype`

but I don't see how it can change edf on an existing results instance given that it is used in results `__init__` and cached. It doesn't when I run an example.

The edf computation depends on model scale through hessian_factor if it is not provided as argument
GLM.hessian_factor has:
```
        if scale is None:
            scale = self.estimate_scale(mu)
```

I changed it so that hessian_factor in edf uses self.scale

josef-pkt commented:
It looks like master is all red and full of deprecation warnings (based on travis for this PR).

Gam in this PR looks fine now, except for lint violations
I have one more amend to the "wrong state" in the model instance plus some uncommitted doc changes.

josef-pkt commented:
I am planning to merge this tomorrow. (Never merge a big PR late at night.)
